cscs-checks/microbenchmarks/gpu/gpu_burn/src/gpu_burn.cu

```diff
@@ -39,553 +42,369 @@
-                    // fflush(stdout);
-                    // system("date"); // Printing a date
-                    // fflush(stdout);
-                    // printf("\n");
-                    // printf("\t(checkpoint)\n");
-                    for (size_t i = 0; i < clientErrors.size(); ++i) {
-                        if (clientErrors.at(i))
-                            clientFaulty.at(i) = true;
-                        clientErrors.at(i) = 0;
-                    }
-                }
-            }
-
-            // Checking whether all clients are dead
-            bool oneAlive = false;
-            for (size_t i = 0; i < clientCalcs.size(); ++i)
-                if (clientCalcs.at(i) != -1)
-                    oneAlive = true;
-            if (!oneAlive) {
-                fprintf(stderr, "\n\nNo clients are alive!  Aborting\n");
-                exit(123);
-            }
-
-            if (startTime + runTime < thisTime)
-                break;
-    }
+    // Here burn is a switch that holds and breaks the work done by the slave threads.
+    burn = false;
 
-    // printf("\nKilling processes.. ");
-    fflush(stdout);
-    for (size_t i = 0; i < clientPid.size(); ++i)
-        kill(clientPid.at(i), 15);
+    XGetDeviceCount(&devCount);
+    std::vector<std::thread> threads;
 
-    // kill(tempPid, 15);
-    // close(tempHandle);
+    // Print device count
+    printf("[%s] Found %d device(s).\n", hostname, devCount);
 
-    while (wait(NULL) != -1);
-    printf("Node %s:\n", hostname);
+    // All the burn info is stored in instances of the BurnTracker class.
+    BurnTracker * trackThreads = new BurnTracker[devCount];
 
-    // printf(" Tested %d GPUs: ", (int)clientPid.size());
-    for (size_t i = 0; i < clientPid.size(); ++i) {
-        printf("  GPU %2d(%s): %4.0f GF/s  %i Celsius\n", (int)i,clientFaulty.at(i) ? "FAULTY" : "OK", clientGflops.at(i), clientTemp.at(i));
+    // Create one thread per device - burn is still off here.
+    for (int i = 0; i < devCount; i++)
+    {
+        BurnTracker * bt = new (&trackThreads[i]) BurnTracker();
+        threads.push_back(std::thread(startBurn<T>,
+                                      i, smi_handle,
+                                      A, B,
+                                      bt
+                          )
+        );
     }
-    printf("\n");
-}
-
-template<class T> void launch(int runLength, bool useDoubles) {
-    //    std::system("nvidia-smi -L");
 
-    // Initializing A and B with random data
-    T *A = (T*) malloc(sizeof(T)*SIZE*SIZE);
-    T *B = (T*) malloc(sizeof(T)*SIZE*SIZE);
-    srand(10);
-    for (size_t i = 0; i < SIZE*SIZE; ++i) {
-        A[i] = (T)((double)(rand()%1000000)/100000.0);
-        B[i] = (T)((double)(rand()%1000000)/100000.0);
+    // Hold until all the threads are done with the init.
+    {
+        std::unique_lock<std::mutex> lk(cv_m);
+        cv.wait(lk, []{return startUpCounter == devCount;});
```

Comment:
Perhaps you can replace those with barriers in C++20.

Reply:
Wouldn't this be a bit early for systems like Daint where GCC there is at 8.2?

Reply:
Yes, leave it as it is now. It's just a comment for the future.